### ISSUE TITLE
Request to Speed up `doUpdateFileContents`

### ISSUE THREAD

**Issue by anagrius** [halohalospecial/atom-elmjutsu#105]
Hello Halo,

I am having performance issues when navigating around my Elm files. The reason looks to be `doUpdateFileContents` in the Indexer. I know it needs to do some work to collect type info etc.
But it seems to trigger a bit too optimistically and on big projects I often get 700ms-2000ms hanging UI.

Are you traversing files too often even when the underlying file set has not changed? Do we need to do it even when only jumping around a file and not saving?

I have included a Timeline dump from Atom's Chrome Debugger. You can try and load it to see what I mean. Also I have a screen shot of what seems to be the buttle neck.

Maybe just making that part more performant will make me happier. I already owe you so much beer, I might as well owe you another.

/Thomas

Timeline Dump (Load it in Chrome):
https://gist.github.com/anagrius/882cd79709321edbede6daadd16a6fa9

Not sure this IS actually the bottle neck:
![screen shot 2017-11-23 at 11 50 34](https://user-images.githubusercontent.com/157210/33169830-5b2205aa-d046-11e7-8b15-396150e19c65.png)

**Comment by halohalospecial** [comment]
Hi @anagrius, thanks for the dump!  I created a new branch (https://github.com/halohalospecial/atom-elmjutsu/tree/perf) that I'll update from time to time.  Can you try that branch out and check if it alleviates the issue?  Thanks!

**Comment by anagrius** [comment]
Super, I have installed the branch. I will have to see over the coming days if it has improved. My measurement will be my gut feeling. I don't know any other easy way to give you feedback :)

**Comment by halohalospecial** [comment]
Thanks!

**Comment by anagrius** [comment]
So, I have been testing it a lot. Seems a lot better (mind you, it might just be placebo!)

Anyway, I did encounter some performance problems again, but they seem to come from other functions. I have provided another timeline dump. Let me know if they are not useful or I need to provide more information in them like memory usage.


[TimelineRawData-20171128T122102.json.zip](https://github.com/halohalospecial/atom-elmjutsu/files/1509659/TimelineRawData-20171128T122102.json.zip)
